aten/src/ATen/core/autocast/register_autocast_ops.cpp

```diff
@@ -0,0 +1,528 @@
+// source tensor in cached_casts is unnecessary but also does no harm (does not increase memory use).
+//
+// When the autocast context manager exits, which should occur at the end of each forward pass, it calls
+// clear_cache to ensure cached Tensors don't leak outside the autocasting region.
+thread_local std::unordered_map<TensorImpl*, std::tuple<Tensor, Tensor>> cached_casts;
+}
+
+void clear_cache() {
+  cached_casts.clear();
+}
+
+enum class CastPolicy : uint8_t {
+  fp16 = 0, // Cast all inputs to fp16 before running the op.
+  fp16_with_tensorlist, // Separate from CastPolicy::fp16 due to lifetime concerns of data
+                        // viewed by TensorLists.  Maybe some C++ guru can help me unify them.
+  fp32, // Cast all inputs to fp32 before running the op.
+  fp32_set_opt_dtype, // fp32_set_opt_dtype and fp32_append_dtype handle functions that should
+  fp32_append_dtype, // output in fp32 and support an output dtype flag.  The policy is:
+                     // If the user has explicity specified a dtype, respect it.
+                     // Otherwise, set the flag (if present in arglist) or append it
+                     // (if not present) to ensure outputs are fp32.
+  promote, // Run in the widest dtype among several args.
+  promote_with_tensorlist, // Separate from CastPolicy::promote due to lifetime concerns
+                           // of data viewed by TensorLists.
+};
+
+/********************************************************************
+Logic to extract the promote type from any Tensor or TensorList args.
+********************************************************************/
+
+// Overload to catch Tensor args.
+// If nextArg is floating-point, compare its scalar_type with our
+// current best guess for the promote type, and update if necessary.
+inline at::ScalarType
+prioritize(at::ScalarType current, const Tensor& nextArg) {
+  if (nextArg.is_cuda() && nextArg.is_floating_point()) {
+    auto next = nextArg.scalar_type();
+    // For promotion purposes, prioritize double, then float, then half.
+    if (current == at::kDouble || next == at::kDouble) {
+      return at::kDouble;
+    } else if (current == at::kFloat || next == at::kFloat) {
+      return at::kFloat;
+    } else if (current == at::kHalf && next == at::kHalf) {
+      return at::kHalf;
+    } else {
+      AT_ERROR("Unexpected floating ScalarType in at::autocast::prioritize");
+      return current;
+    }
+  } else {
+    return current;
+  }
+}
+
+// Overload to catch TensorList args (for e.g. cat, stack).
+// Reuses the overload above to process each Tensor in the list.
+inline at::ScalarType
+prioritize(at::ScalarType current, const TensorList& list) {
+  for (const auto& tensor : list) {
+    current = prioritize(current, tensor);
+  }
+  return current;
+}
+
+// Template to catch non-Tensor args (no-op that returns current best guess)
+template<typename T> inline at::ScalarType
+prioritize(at::ScalarType current, T nextArg) {
```

Comment:
cc @izdeby as this is somewhat related to TensorOptions removal (I don't think any action is needed, but I think you should be aware of this new code that is doing matching for ScalarType arguments)